src/network/networkd-sysctl.c

```diff
@@ -5,17 +5,189 @@
+                return log_error_errno(errno, "Failed to update cgroup map: %m");
+
+        sysctl_link = sym_bpf_program__attach_cgroup(obj->progs.sysctl_monitor, rootcg);
+        r = bpf_get_error_translated(sysctl_link);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to attach sysctl monitor BPF program to cgroup: %m");
+
+        sysctl_buffer = sym_ring_buffer__new(
+                        sym_bpf_map__fd(obj->maps.written_sysctls),
+                        sysct_write_event, &manager->sysctl_shadow, NULL);
+        if (!sysctl_buffer)
+                return log_warning_errno(errno, "Failed to create ring buffer: %m");
+
+        r = sd_event_add_io(manager->event, &manager->sysctl_event_source,
+                        sym_ring_buffer__epoll_fd(sysctl_buffer), EPOLLIN, on_ringbuf_io, sysctl_buffer);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to watch sysctl event ringbuffer: %m");
+
+        manager->sysctl_link = TAKE_PTR(sysctl_link);
+        manager->sysctl_skel = TAKE_PTR(obj);
+        manager->sysctl_buffer = TAKE_PTR(sysctl_buffer);
+        manager->cgroup_fd = TAKE_FD(cgroup_fd);
+
+        return 0;
+}
+
+int sysctl_remove_monitor(Manager *manager) {
+        _cleanup_close_ int rootcg = -EBADF;
+        int r;
+
+        assert(manager);
+
+        r = dlopen_bpf();
+        if (r < 0)
+                return r;
+
+        rootcg = cg_path_open(SYSTEMD_CGROUP_CONTROLLER, "/");
+        if (rootcg < 0)
+                return r;
+
+        if (manager->sysctl_event_source)
+                manager->sysctl_event_source = sd_event_source_disable_unref(manager->sysctl_event_source);
+
+        if (manager->sysctl_buffer) {
+                sym_ring_buffer__free(manager->sysctl_buffer);
+                manager->sysctl_buffer = NULL;
+        }
+
+        if (manager->sysctl_link) {
+                sym_bpf_link__destroy(manager->sysctl_link);
+                manager->sysctl_link = NULL;
+        }
+
+        if (manager->sysctl_skel) {
+                sysctl_monitor_bpf__destroy(manager->sysctl_skel);
+                manager->sysctl_skel = NULL;
+        }
+
+        if (manager->cgroup_fd >= 0)
+                manager->cgroup_fd = safe_close(manager->cgroup_fd);
+
+        return 0;
+}
+
+int sysctl_clear_link_shadows(Link *link) {
+        _cleanup_free_ char *ipv4 = NULL, *ipv6 = NULL;
+        char *key = NULL, *value = NULL;
+
+        assert(link);
+        assert(link->manager);
+
+        ipv4 = path_join("/proc/sys/net/ipv4/conf", link->ifname);
+        if (!ipv4)
+                return log_oom();
+
+        ipv6 = path_join("/proc/sys/net/ipv6/conf", link->ifname);
+        if (!ipv6)
+                return log_oom();
+
+        HASHMAP_FOREACH_KEY(value, key, link->manager->sysctl_shadow)
+                if (path_startswith(key, ipv4) || path_startswith(key, ipv6)) {
+                        hashmap_remove(link->manager->sysctl_shadow, key);
```

Comment:
maybe even:

```c
assert_se(hashmap_remove(…, key) == value);
```

simply to encode the expectation here.

Reply:
hashmap_remove_value suits even better.

Reply:
I quickly looked at the `hashmap_remove_value()` source, is it a `hashmap_remove()` which conditionally remove the entry only if the supplied value matches the one in the hashmap?
If so, how could be different than the value we just found by iterating?

Reply:
Well, it can't be, hence the assert_se.